orca/producer.go

```diff
@@ -0,0 +1,219 @@
+		intervals: make(map[time.Duration]int),
+		listeners: make(map[OOBListener]struct{}),
+		backoff:   internal.DefaultBackoffFunc,
+	}
+	go p.run(ctx)
+	return p, func() {
+		cancel()
+		<-p.closed.Done() // Block until stream stopped.
+	}
+}
+
+var producerBuilderSingleton = &producerBuilder{}
+
+// OOBListener is used to receive out-of-band load reports as they arrive.
+type OOBListener interface {
+	// OnLoadReport is called when a load report is received.
+	OnLoadReport(*v3orcapb.OrcaLoadReport)
+}
+
+// OOBListenerOptions contains options to control how an OOBListener is called.
+type OOBListenerOptions struct {
+	// How often to request the server to provide a load report.  May be
+	// provided less frequently if the server requires a longer interval, or
+	// may be provided more frequently if another subscriber requests a shorter
+	// interval.
+	ReportInterval time.Duration
+}
+
+// RegisterOOBListener registers an out-of-band load report listener on sc.
+// Any OOBListener may only be registered once per subchannel at a time.  The
+// returned stop function must be called when no longer needed.
+func RegisterOOBListener(sc balancer.SubConn, l OOBListener, opts OOBListenerOptions) (stop func()) {
+	pr, close := sc.GetOrBuildProducer(producerBuilderSingleton)
+	p := pr.(*producer)
+	p.registerListener(l, opts.ReportInterval)
+
+	// TODO: When we can register for SubConn state updates, don't call run()
+	// until READY and automatically call stop() on SHUTDOWN.
+
+	// If stop is called multiple times, prevent it from having any effect on
+	// subsequent calls.
+	var once sync.Once
+	return func() {
+		once.Do(func() {
+			p.unregisterListener(l, opts.ReportInterval)
+			close()
+		})
+	}
+}
+
+type producer struct {
+	client v3orcaservicegrpc.OpenRcaServiceClient
+
+	ctx     context.Context
+	cancel  func()
+	closed  *grpcsync.Event // fired when closure completes
+	backoff func(int) time.Duration
+
+	mu        sync.Mutex
+	intervals map[time.Duration]int    // map from interval time to count of listeners requesting that time
+	listeners map[OOBListener]struct{} // set of registered listeners
+}
+
+// registerListener adds the listener and its requested report interval to the
+// producer.
+func (p *producer) registerListener(l OOBListener, interval time.Duration) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.listeners[l] = struct{}{}
+	p.intervals[interval]++
+}
+
+// registerListener removes the listener and its requested report interval to
+// the producer.
+func (p *producer) unregisterListener(l OOBListener, interval time.Duration) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	delete(p.listeners, l)
+	p.intervals[interval]--
+	if p.intervals[interval] == 0 {
+		delete(p.intervals, interval)
+	}
+}
+
+// minInterval returns the smallest key in p.intervals.
+func (p *producer) minInterval() time.Duration {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	min := time.Duration(-1)
+	for t := range p.intervals {
+		if min == time.Duration(-1) || t < min {
```

Comment:
Invariant is if done ctx.Err() returns, then it all exits?

Reply:
> Invariant is if done ctx.Err() returns, then it all exits?

Yes.  When the subscribers (`== len(intervals)`) have all closed then the context is canceled and this will never be called again.

> If so, why not just initialize min to the zero value (I'm assuming a listener can't request a 0 time interval?)

It _should_ be okay to use `0` instead, but if any consumer does use `0` then the following value would be selected.  I'll remove the magic entirely (it was always a bad way of doing this), just in case someone does something unexpected.

Reply:
Thanks. Both what you had originally and now cause "but if any consumer does use 0 then the following value would be selected". I know it shouldn't happen regardless now (or the -1 returned) because of the invariant protecting, but what happens if ORCA is called with a 0 time interval in OrcaLoadReportRequest.

Reply:
It's fine.  Did you see the updated code?  It is more robust than the previous method.

Reply:
Yeah, thanks for updating. Just left this comment open about the 0 report time interval.